Feature/git docs and 3rdparty (#5)
joinbox/guidelines: styleguide/git/branching-model.md

# Git Branching Model

Joinbox uses a more specific version of the git-flow branching model

![Alt text](git-flow.png "Joinbox GIT Flow")

> ## General

Reviewer: Missing a bullet point which explains that all feature branches are only merged after a successful review.

Author: added

- the master branch can be released at any time, it is always 100% stable
- the develop branch contains only code that works and is stable but not yet tested by the client
- features and bug fixes are developed in feature branches until they are stable. the base branch for creating feature branches is the release branch where the issue is in or the master branch for hot fixes.

> - feature branches may only merged into the develop branch if they are tested and work correctly

Reviewer: may only **be** merged

- for each iteration (release cycle, milestone) a new release branch is created from develop branch when the development starts.
- the release branch is published to a testing server for acceptance tests by the client. the client should receive the change log of the release branch
- fixes for the release branch are merged into the release branch
- the release branch may be merged into develop or another release branch if it is required
- hot fixes are only created if they must be applied to the master branch for immediate release. they are based on the master branch

> - after a hot fix or release was released the master branch is merged into the develop branch

Reviewer: Shouldn't we merge it into the current release branch if there exists one?

> - releases must be planned in advance, there should no more than one release branch at any time

Reviewer: there should **not be** more

- after the release branch was merged into master and tagged with the correct version the release branch may be deleted.

## Naming

> Branches must not be updated after a review was requested, before changes are requested in the review.

Reviewer: I already asked this: why not?

Author: I changed this form after opening a PR to after requesting a review form your suggestion.
The Idea behind is to avoid people using time to review code an the the developer updates the code because it was not finished so one needs to review the code again. So basically a way to tell when one need to make a review.

# Git Branching Model

Joinbox uses a more specific version of the git-flow branching model

![Joinbox Git Flow Model](git-flow.png "Joinbox Git Flow")

## General

- the master branch can be released at any time, it is always 100% stable

> - the develop branch contains only code that works and is stable but not yet tested by the client

Reviewer 2: ```suggestion
- The develop branch contains only code that works and is stable but is not yet tested by the client.
```

> - features and bug fixes are developed in feature branches until they are stable. the base branch for creating feature branches is the release branch where the issue is in or the master branch for hot fixes.

Reviewer 2: ```suggestion
- Features and bug fixes are developed in feature branches until they are stable. The branch, for basing feature branches on, is the release branch to which the task belongs to, or in the case of hot-fixes, the master branch.
```

- feature branches may only be merged into the release branch if they are tested and work correctly.
- feature branches will only be merged in to the according branch after at least one reviewer approved the pull request.

> - for each iteration (release cycle, milestone) a new release branch is created from develop branch when the development starts.

Reviewer 2: this should be above the `- features and bug fixes are developed in feature [...]` item since it's referenced there

> - the release branch is published to a testing server for acceptance tests by the client. the client should receive the change log of the release branch

Reviewer 2: this is inconsistent with `the develop branch contains only code that works and is stable but not yet tested by the client`

- fixes for the release branch are merged into the release branch
- the release branch may be merged into develop or another release branch if it is required
- hot fixes are only created if they must be applied to the master branch for immediate release. they are based on the master branch

> - after a hot fix or release was released the master branch is merged into the develop branch as well as open release branches

Reviewer 2: who does that? the owner of the branch?

> - releases must be planned in advance, there should not be more than one release branch at any time

Reviewer 2: this may not be doable. this should be a recommendation.

> - after the release branch was merged into master and tagged with the correct version the release branch may be deleted.

Reviewer 2: dont mention this here `and tagged with the correct version`. this is a separate item: `If the project has version numbers, the correct version number must be set after the release branch was merged into the master branch.`

Author: Do you think we should make it optional to use version numbers for a project?

## Naming

> Branches must not be updated after a review was requested, before changes are requested in the review. If the developer want's to track the progress, a pull request without a reviewer assigned may be opened.

Reviewer 2: ```suggestion
If a review on a pull request is requested, the corresponding branch must not be updated until changes were requested by the reviewer. If the developer wants to track the progress, a pull request without a reviewer assigned may be opened.
```

Reviewer 2: What does `If the developer wants to track the progress, a pull request without a reviewer assigned may be opened` mean? What's the use case?

Author: @rcky uses that and suggested to add the possibility to do so.

> - A feature branch must be created for each issue based on the release branch. The name must follow the pattern `feature/iss-<Issue-Number>-<Short-Description-of-the-Issue>`. There may be multiple feature branches for one issue.

Reviewer 2: Please don't make this issue centric, it's a generic guideline. How about tasks instead of issues? That's a bit less specific. You may add some example names for issues though.

- A release branch must be created for each milestone based on the current develop branch. The name must follow the pattern `release/<Milestone-Name>`

> A fix branch may be created for bugs or forgotten parts of a planed issue ( inside a milestone ) based on the release branch. The name must follow the pattern `fix/iss-<Issue-Number>-<Short-Description-of-the-Issue>`

Reviewer 2: ```suggestion
A fix branch may be created for bugs or forgotten parts of a planned issue (inside a milestone) based on the release branch. The name must follow the pattern `fix/iss-<Issue-Number>-<Short-Description-of-the-Issue>`
```
